src/llvm_ir_generator.cpp

```diff
@@ -0,0 +1,807 @@
+  switch (op) {
+    case BinaryOperator::kGt:
+    case BinaryOperator::kGte:
+    case BinaryOperator::kLt:
+    case BinaryOperator::kLte:
+    case BinaryOperator::kEq:
+    case BinaryOperator::kNeq:
+      return true;
+    default:
+      return false;
+  }
+}
+
+auto id_to_val  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables):
+                // Accessible only within this translation unit; declaring as a
+                // data member introduces unnecessary dependency.
+    = std::map<std::string, llvm::Value*>{};
+
+auto
+    val_to_id_addr  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables):
+                    // Accessible only within this translation unit; declaring
+                    // as a data member introduces unnecessary dependency.
+    = std::map<llvm::Value*, llvm::Value*>{};
+
+/// @brief Every expression generates a LLVM object that is the subclass of
+/// `llvm::Value`. This object is stored, so we can propagate to later use.
+class PrevValueRecorder {
+ public:
+  void Record(llvm::Value* val) {
+    prev_val_ = val;
+  }
+
+  llvm::Value* ValOfPrevExpr() {
+    assert(prev_val_);
+    return prev_val_;
+  }
+
+ private:
+  llvm::Value* prev_val_ = nullptr;
+};
+
+auto
+    val_recorder  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables):
+                  // Accessible only within this translation unit; declaring as
+                  // a data member introduces unnecessary dependency.
+    = PrevValueRecorder{};
+
+struct LabelViewInfo {
+  llvm::BasicBlock* entry;
+  llvm::BasicBlock* exit;
+  /// @brief This vector stores every `case` and `default` basic blocks of
+  /// a switch case.
+  /// This first element of a pair is the expression value
+  /// of a case statement.
+  /// This second element of a pair is the label's basic block.
+  std::vector<std::pair<llvm::Value*, llvm::BasicBlock*>> cases{};
+};
+
+/// @note Blocks that allows jumping within or out of it should add its labels
+/// to this list.
+auto
+    label_views_of_jumpable_blocks  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+    = std::vector<LabelViewInfo>{};
+
+}  // namespace
+
+void LLVMIRGenerator::Visit(const DeclStmtNode& decl_stmt) {
+  for (const auto& decl : decl_stmt.decls) {
+    decl->Accept(*this);
+  }
+}
+
+void LLVMIRGenerator::Visit(const VarDeclNode& decl) {
+  auto var_type = llvm_util_.GetLLVMType(*(decl.type));
+  // For function pointer, we need to change from FunctionType to PointerType
+  var_type = var_type->isFunctionTy() ? var_type->getPointerTo() : var_type;
+  auto addr = builder_->CreateAlloca(var_type);
+  if (decl.init) {
+    decl.init->Accept(*this);
+    auto val = val_recorder.ValOfPrevExpr();
+    builder_->CreateStore(val, addr);
+  }
+  id_to_val[decl.id] = addr;
+}
+
+void LLVMIRGenerator::Visit(const ArrDeclNode& arr_decl) {
+  auto arr_type = llvm_util_.GetLLVMType(*(arr_decl.type));
+  auto base_addr = builder_->CreateAlloca(arr_type, nullptr);
```

Comment:
The default value is already `nullptr`. We can omit it.
```suggestion
  auto base_addr = builder_->CreateAlloca(arr_type);
```